minicluster/src/main/java/org/apache/accumulo/miniclusterImpl/MiniAccumuloClusterImpl.java

```diff
@@ -824,9 +825,46 @@ public synchronized void stop() throws IOException, InterruptedException {
 
     control.stop(ServerType.GARBAGE_COLLECTOR, null);
     control.stop(ServerType.MANAGER, null);
+    control.stop(ServerType.COMPACTION_COORDINATOR);
     control.stop(ServerType.TABLET_SERVER, null);
+    control.stop(ServerType.COMPACTOR, null);
+    control.stop(ServerType.SCAN_SERVER, null);
+
+    // The method calls above kill the server
+    // Clean up the locks in ZooKeeper fo that if the cluster
+    // is restarted, then the processes will start right away
+    // and not wait for the old locks to be cleaned up.
+    try {
+      new ZooZap().zap(getServerContext().getSiteConfiguration(), new String[] {"-manager",
+          "-compaction-coordinators", "-tservers", "-compactors", "-sservers"});
+    } catch (Exception e) {
+      log.error("Error zapping zookeeper locks", e);
+    }
     control.stop(ServerType.ZOOKEEPER, null);
 
+    // Clear the location of the servers in ZooCache.
+    // When ZooKeeper was stopped in the previous method call,
+    // the local ZooKeeper watcher did not fire. If MAC is
+    // restarted, then ZooKeeper will start on the same port with
+    // the same data, but no Watchers will fire.
+    boolean startCalled = true;
+    try {
+      getServerContext();
+    } catch (RuntimeException e) {
+      if (e.getMessage().startsWith("Accumulo not initialized")) {
+        startCalled = false;
+      }
+    }
+    if (startCalled) {
+      final ServerContext ctx = getServerContext();
+      final String zRoot = getServerContext().getZooKeeperRoot();
+      ctx.getZooCache().clear((path) -> path.startsWith(zRoot + Constants.ZMANAGER_LOCK));
+      ctx.getZooCache().clear((path) -> path.startsWith(zRoot + Constants.ZGC_LOCK));
+      ctx.getZooCache().clear((path) -> path.startsWith(zRoot + Constants.ZCOMPACTORS));
+      ctx.getZooCache().clear((path) -> path.startsWith(zRoot + Constants.ZSSERVERS));
+      ctx.getZooCache().clear((path) -> path.startsWith(zRoot + Constants.ZTSERVERS));
```

Comment:
I don't know if this is helpful, but you can define the predicate once outside this code, and then just call clear a single time with a single ready-to-go predicate.

```suggestion
      Predicate<String> lockPaths = path -> path.startsWith(zRoot + Constants.ZMANAGER_LOCK);
      for (String s : Set.of(Constants.ZGC_LOCK, Constants.ZCOMPACTORS, Constants.ZSSERVERS, Constants.ZTSERVERS)) {
        lockPaths = lockPaths.or(path -> path.startsWith(zRoot + s));
      }
      ctx.getZooCache().clear(lockPaths);
```

You could also define it a predicate that is true if any predicate on a stream is true:

```suggestion
      Predicate<String> lockPaths = path -> Stream.of(Constants.ZMANAGER_LOCK, Constants.ZGC_LOCK, Constants.ZCOMPACTORS, Constants.ZSSERVERS, Constants.ZTSERVERS)
          .anyMatch(p -> path.startsWith(zRoot + p));
      ctx.getZooCache().clear(lockPaths);
```

The setup of the Predicate is a little more costly, but if done once per jvm as a static final, it could be worth it to reduce the number of calls to zooCache.clear(), and it might be a little more readable.

Reply:
Wouldn't this variant test each key in the cache against `Stream.of(Constants.ZMANAGER_LOCK, Constants.ZGC_LOCK, Constants.ZCOMPACTORS, Constants.ZSSERVERS, Constants.ZTSERVERS).anyMatch(p -> path.startsWith(zRoot + p))` ? Wouldn't it create the Stream for every key?

Reply:
> Wouldn't it create the Stream for every key?

Yeah, I would avoid the second variant. I was just offering inspiration for a possibly more readable version. That second one would perform badly, though. The first variant is slightly less readable, but would perform much better, as there's only a one time cost to set up the chained predicate, especially if the predicate is saved outside the method in some static final.